core/src/test/java/org/apache/iceberg/TestRowLineageMetadata.java

```diff
@@ -0,0 +1,328 @@
+    // previous metadata and not
+    // change last-row-id for this metadata.
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(30);
+    assertThat(table.currentSnapshot().addedRows()).isEqualTo(0);
+    assertThat(table.ops().current().nextRowId()).isEqualTo(30);
+  }
+
+  @TestTemplate
+  public void testReplace() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    TestTables.TestTable table =
+        TestTables.create(
+            tableDir, "test", TEST_SCHEMA, PartitionSpec.unpartitioned(), formatVersion);
+    TableMetadata base = table.ops().current();
+
+    table.ops().commit(base, TableMetadata.buildFrom(base).enableRowLineage().build());
+
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+    assertThat(table.ops().current().nextRowId()).isEqualTo(0L);
+
+    DataFile filePart1 = fileWithRows(30);
+    DataFile filePart2 = fileWithRows(30);
+    DataFile fileCompacted = fileWithRows(60);
+
+    table.newAppend().appendFile(filePart1).appendFile(filePart2).commit();
+
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(0);
+    assertThat(table.ops().current().nextRowId()).isEqualTo(60);
+
+    table.newRewrite().deleteFile(filePart1).deleteFile(filePart2).addFile(fileCompacted).commit();
+
+    // Rewrites are currently just treated as appends. In the future we could treat these as no-ops
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(60);
+    assertThat(table.ops().current().nextRowId()).isEqualTo(120);
+  }
+
+  @TestTemplate
+  public void testEnableRowLineageViaProperty() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    TestTables.TestTable table =
+        TestTables.create(
+            tableDir, "test", TEST_SCHEMA, PartitionSpec.unpartitioned(), formatVersion);
+
+    assertThat(table.ops().current().rowLineageEnabled()).isFalse();
+
+    // No-op
+    table.updateProperties().set(TableProperties.ROW_LINEAGE, "false").commit();
+    assertThat(table.ops().current().rowLineageEnabled()).isFalse();
+
+    // Enable row lineage
+    table.updateProperties().set(TableProperties.ROW_LINEAGE, "true").commit();
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+
+    // Disabling row lineage is not allowed
+    assertThatThrownBy(
+            () -> table.updateProperties().set(TableProperties.ROW_LINEAGE, "false").commit())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Cannot disable row lineage once it has been enabled");
+
+    // No-op
+    table.updateProperties().set(TableProperties.ROW_LINEAGE, "true").commit();
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
```

Comment:
Nit: I don't feel like this is strictly neccessary?

Reply:
This one I think we want, since we are making sure that the operation is actually a no-op